modules/contact/src/userobjects/WeightedGapUserObject.C

```diff
@@ -0,0 +1,155 @@
+  _test = &test();
+}
+
+void
+WeightedGapUserObject::computeQpProperties()
+{
+  // Trim interior node variable derivatives
+  const auto & primary_ip_lowerd_map = amg().getPrimaryIpToLowerElementMap(
+      *_lower_primary_elem, *_lower_primary_elem->interior_parent(), *_lower_secondary_elem);
+  const auto & secondary_ip_lowerd_map =
+      amg().getSecondaryIpToLowerElementMap(*_lower_secondary_elem);
+
+  std::array<const MooseVariable *, 3> var_array{{_disp_x_var, _disp_y_var, _disp_z_var}};
+  std::array<ADReal, 3> primary_disp{
+      {_primary_disp_x[_qp], _primary_disp_y[_qp], _has_disp_z ? (*_primary_disp_z)[_qp] : 0}};
+  std::array<ADReal, 3> secondary_disp{{_secondary_disp_x[_qp],
+                                        _secondary_disp_y[_qp],
+                                        _has_disp_z ? (*_secondary_disp_z)[_qp] : 0}};
+
+  trimInteriorNodeDerivatives(primary_ip_lowerd_map, var_array, primary_disp, false);
+  trimInteriorNodeDerivatives(secondary_ip_lowerd_map, var_array, secondary_disp, true);
+
+  const ADReal & prim_x = primary_disp[0];
+  const ADReal & prim_y = primary_disp[1];
+  const ADReal * prim_z = nullptr;
+  if (_has_disp_z)
+    prim_z = &primary_disp[2];
+
+  const ADReal & sec_x = secondary_disp[0];
+  const ADReal & sec_y = secondary_disp[1];
+  const ADReal * sec_z = nullptr;
+  if (_has_disp_z)
+    sec_z = &secondary_disp[2];
+
+  // Compute gap vector
+  ADRealVectorValue gap_vec = _phys_points_primary[_qp] - _phys_points_secondary[_qp];
+
+  gap_vec(0).derivatives() = prim_x.derivatives() - sec_x.derivatives();
+  gap_vec(1).derivatives() = prim_y.derivatives() - sec_y.derivatives();
+  if (_has_disp_z)
+    gap_vec(2).derivatives() = prim_z->derivatives() - sec_z->derivatives();
+
+  // Compute integration point quantities: Normals (geometry) is averaged at the node, but not
+  // interpolated within the weak integration.
+  _qp_gap_nodal = gap_vec * (_JxW_msm[_qp] * _coord[_qp]);
+
+  // To do normalization of constraint coefficient (c_n)
+  _qp_factor = _JxW_msm[_qp] * _coord[_qp];
+}
+
+void
+WeightedGapUserObject::computeQpIProperties()
+{
+  mooseAssert(_normals.size() == _lower_secondary_elem->n_nodes(),
+              "Making sure that _normals is the expected size");
+
+  // Get the _dof_to_weighted_gap map
+  const auto * const dof = static_cast<const DofObject *>(_lower_secondary_elem->node_ptr(_i));
+
+  _dof_to_weighted_gap[dof].first += (*_test)[_i][_qp] * _qp_gap_nodal * _normals[_i];
+
+  if (_normalize_c)
+    _dof_to_weighted_gap[dof].second += (*_test)[_i][_qp] * _qp_factor;
+}
+
+void
+WeightedGapUserObject::initialize()
+{
+  _dof_to_weighted_gap.clear();
+}
+
+void
+WeightedGapUserObject::finalize()
+{
+  // If the constraint is performed by the owner, then we don't need any data sent back; the owner
+  // will take care of it. But if the constraint is not performed by the owner and we might have to
+  // do some of the constraining ourselves, then we need data sent back to us
```

Comment:
This is not clear to me. Why would anyone but the owner perform the constraining?

Reply:
The penalty constraint is a distributed constraint. You may compute a weighted gap for Node 0 owned by processor 0, and that node is attached to an element on processor 0 and another element on processor 1. The penalty constraint driving the weighted gap on node 0 to zero will be applied through the contact force applied to both the element on processor 0 and the element on processor 1. Whereas for a Lagrange multiplier implementation the constraint is enforced in the Lagrange multiplier constraint object itself and only on a local basis